### DISCUSSION TITLE
Possible to extract .NET expression trees and use them standalone in my code?

### DISCUSSION THREAD

**Discussion by maxima120** [sebastienros/jint#1529] [Q&A]
I understand that jint is parsing js script into .NET expression trees. Is that correct? 
I would like to have access to it in order to use in a project where calculations will be ran twice a second on thousands of objects. I want full control of it. So the natural way to have it in a manageable / maintainable fashion is to use jint to parse scripts initially into expression trees (several dozens of them), compile and then run repeatedly.. Is this something I can do now? If not - can it be a feature request and in that case what would be expected time to implement, if accepted? 

And if all fails - I would probably need to fork it. What class should I be looking into first in order to get parsed expression trees?

**Comment by Genteure** [comment]
> I understand that jint is parsing js script into .NET expression trees. Is that correct?

No, Jint does not parse javascript into `System.Linq.Expressions.Expression`. Jint is a javascript **interpreter**, it does not compile anything.

> ... calculations will be ran twice a second on thousands of objects. I want full control of it. 

Sounds like a reasonable job for Jint as is, did you encounter any specific problems? Can you elaborate how exactly do you want to _control it_?

> ... parse scripts initially into expression trees (several dozens of them), compile and then run repeatedly ...

You can call `Engine.PrepareScript` and `Engine.PrepareModule` to parse `string` into `Script` and `Module` to avoid re-parsing every time, but Jint does not compile anything.

**Comment by maxima120** [comment]
> Jint is a javascript interpreter, it does not compile anything.

how does it execute the script (a very general description please)? You must convert it into computable entities somehow (since you do not use any external process, see here: https://github.com/sebastienros/jint/discussions/1526). A stack with operations and operands?

I saw use of Expression dotnet object in `JintExpression` class this is why I concluded you parse it into expression tree. Would it be hard to convert `JintExpression` into expression tree?

**Reply by Genteure** [reply]
It's impossible to convert `Jint.Runtime.Interpreter.Expressions.JintExpression` into `System.Linq.Expressions.Expression`.
The `Expression` in `JintExpression` is `Esprima.Ast.Expression`, representing javascript AST, not `System.Linq.Expressions.Expression`, the only similarity between them is their name.

> how does it execute the script

Maybe the Wikipedia article [Interpreter (computing)](https://en.wikipedia.org/wiki/Interpreter_(computing)) can explain better than I do, but the idea is Jint (technically [Esprima .NET](https://github.com/sebastienros/esprima-dotnet)) first parse the script text (aka `string`) into AST (abstract syntax tree, basically it represent the same thing as the script text, but in .NET object form), then it goes over each element in the AST and do things.

For example, when running `Math.floor(1.2)`, it first finds the `Math` object from the global object, then finds the `floor` object in `Math`, then calls `floor` with the value `1.2`.
The `floor` function is written in C# [here](https://github.com/sebastienros/jint/blob/8b0da8ad3a754b4bb5d7da9d2492acef7ab25aa4/Jint/Native/Math/MathInstance.cs#L463), and added to `Math` [here](https://github.com/sebastienros/jint/blob/8b0da8ad3a754b4bb5d7da9d2492acef7ab25aa4/Jint/Native/Math/MathInstance.cs#L37).

You could also clone the source code, open it with Visual Studio, enter debug mode and step over the process with the debugger.

While it can be helpful to know how Jint work on a lower level, you don't _need_ to understand it to use it. Just like how you don't need to know how C# compiler work or how the .NET Runtime work to write C#, or you don't need to know how electricity flow through transistors to use a computer.

If you really want to use `System.Linq.Expressions.Expression`, only need basic arithmetic operations, want even faster performance, you could try [StringToExpression](https://github.com/codecutout/StringToExpression). You could even write your own string to expression parser/compiler using [parlot](https://github.com/sebastienros/parlot) (also created by `@sebastienros`).
